src/dynamic-launcher.c

```diff
@@ -0,0 +1,794 @@
+                                                     arg_desktop_entry,
+                                                     launcher_data,
+                                                     call->app_info,
+                                                     &icon_path,
+                                                     &error);
+  if (desktop_keyfile == NULL)
+    {
+      g_dbus_method_invocation_return_gerror (invocation, error);
+      return TRUE;
+    }
+
+  /* Put the desktop file in ~/.local/share/xdg-desktop-portal/applications/ so
+   * there's no ambiguity about which launchers were created by this portal.
+   */
+  desktop_dir = g_build_filename (g_get_user_data_dir (), "xdg-desktop-portal", "applications", NULL);
+  g_mkdir_with_parents (desktop_dir, 0755);
+  desktop_path = g_build_filename (desktop_dir, arg_desktop_file_id, NULL);
+  if (!g_key_file_save_to_file (desktop_keyfile, desktop_path, &error))
+    goto error;
+
+  /* Make a sym link in ~/.local/share/applications so the launcher shows up in
+   * the desktop environment's menu.
+   */
+  link_path = g_build_filename (g_get_user_data_dir (), "applications", arg_desktop_file_id, NULL);
+  link_file = g_file_new_for_path (link_path);
+  relative_path = g_build_filename ("..", "xdg-desktop-portal", "applications", arg_desktop_file_id, NULL);
+  g_file_delete (link_file, NULL, NULL);
+  if (!g_file_make_symbolic_link (link_file, relative_path, NULL, &error))
+    goto error;
+
+  result = 0;
+  xdp_dynamic_launcher_complete_install (object, invocation, result);
+  return TRUE;
+
+error:
+  g_dbus_method_invocation_return_gerror (invocation, error);
+  remove (icon_path);
+  remove (desktop_path);
+  remove (link_path);
+  return TRUE;
+}
+
+static XdpOptionKey response_options[] = {
+  { "name", G_VARIANT_TYPE_STRING, NULL },
+  { "token", G_VARIANT_TYPE_UINT32, NULL }
+};
+
+static gboolean
+install_token_timeout (gpointer data)
+{
+  g_autoptr(GVariant) launcher_data = NULL;
+
+  g_debug ("Revoking install token %s", (char *)data);
+  launcher_data = get_launcher_data_and_revoke_token ((char *)data);
+  g_free (data);
+
+  return G_SOURCE_REMOVE;
+}
+
+static void
+set_launcher_data_for_token (const char *token,
+                             GVariant   *launcher_data)
+{
+  g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&transient_permissions_lock);
+
+  if (!transient_permissions)
+    {
+      transient_permissions = g_hash_table_new_full (g_str_hash, g_str_equal,
+                                                     g_free, (GDestroyNotify)g_variant_unref);
+    }
+
+  g_hash_table_insert (transient_permissions,
+                       g_strdup (token),
+                       g_variant_ref (launcher_data));
+
+  /* Revoke the token if it hasn't been used after 5 minutes, in case of
+   * client bugs. This is what the gnome print portal implementation does.
+   */
+  g_timeout_add_seconds (300, install_token_timeout, g_strdup (token));
```

Comment:
You need to store the source id somewhere, and remove the timeout when the token is used